src/libs/decoders/mp3.cpp

```diff
@@ -0,0 +1,222 @@
+/**
+ * This DOSBox mp3 decooder backend is maintained by Kevin R. Croft (krcroft@gmail.com)
+ * This decoder makes use of the following single-header public libraries:
+ *   - dr_mp3: http://mackron.github.io/dr_mp3.html, by David Reid
+ *
+ * The upstream SDL2 Sound 1.9.x mp3 decoder is written and copyright by Ryan C. Gordon. (icculus@icculus.org)
+ *
+ *  This SDL_sound MP3 decoder backend is free software: you can redistribute
+ *  it and/or modify it under the terms of the GNU General Public License as
+ *  published by the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This MP3 decoder backend is distributed in the hope that it
+ *  will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
+ *  of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with the SDL Sound Library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif
+
+#include <assert.h>
+
+#include <SDL.h> // provides: SDL_malloc, SDL_realloc, SDL_free, SDL_memcpy, and SDL_memset
+#define DR_MP3_IMPLEMENTATION
+#define DR_MP3_NO_STDIO 1
+#define DRMP3_ASSERT(x) assert((x))
+#define DRMP3_MALLOC(sz) SDL_malloc((sz))
+#define DRMP3_REALLOC(p, sz) SDL_realloc((p), (sz))
+#define DRMP3_FREE(p) SDL_free((p))
+#define DRMP3_COPY_MEMORY(dst, src, sz) SDL_memcpy((dst), (src), (sz))
+#define DRMP3_ZERO_MEMORY(p, sz) SDL_memset((p), 0, (sz))
+#include "dr_mp3.h" // provides: drmp3
+
+#include "mp3_seek_table.h" // provides: populate_seek_table and SDL_Sound headers
+
+#include "SDL_sound.h"
+#define __SDL_SOUND_INTERNAL__
+#include "SDL_sound_internal.h" // provides: Sound_SampleInternal
+
+#define MP3_FAST_SEEK_FILENAME "fastseek.lut"
+
+static size_t mp3_read(void* const pUserData, void* const pBufferOut, const size_t bytesToRead)
```

Comment:
Hey, aside of using consts and C++ casting, this function is identival to `flac_read`… Maybe these functions should not be static, but an implementation could be reused (maybe by declaring new function in SDL_sound_internal.h or something?).

*This is not a request for change, just a question.*

Reply:
It's an interesting observation! These IO functions need to be passed into these particular codec creation functions, but only one implementation is needed (atleast for these two dr_ codecs). Because they're not SDL-specific, they could go in a generic `drlibs_IO.c` file, or something like that.. hmm. 

For now, perhaps they are tiny enough, and not duplicated enough that would warrant this "genericizing"; plus they are specific to these two DR codecs.  If it was across the board, something that SDL was imposing, then I think _'yes, lets rip them out_' would be a nice improvement.

Reply:
(code left as-is)